### DISCUSSION TITLE
Does cpptools really need to parse over 4 million files?

### DISCUSSION THREAD

**Discussion by whatdoidonow** [microsoft/vscode-cpptools#7618] [Q&A]
I am aware there are already several performance related issues open. But I came across this page https://github.com/microsoft/vscode-cpptools/wiki/Troubleshooting-Performance-Issues which invites you to post performance details in an issue.
Basically, cpptools appears to run continuously, maxing out the core it runs on, and when I hover over the db icon, I see Discovering files: 4164330 (that number is on the rise). It is quite likely that my C++ extension is misconfigured, but I don't know where else to turn to. As far as I'm aware, I am using default settings and the proverbial "I didn't change anything" applies.
As instructed, I shall attach the screenshot of the performance report. That's for C/C++ ms-vscode.cpptools v1.4.0.
![Screenshot from 2021-05-28 13-16-24](https://user-images.githubusercontent.com/57625281/119982799-325fed80-bfb7-11eb-9edb-61d45ab2e445.png)

**Comment by sean-mcmanus** [comment]
It could be following a symlink to your root drive or a folder with a lot of files. You should be able to temporarily set C_Cpp.loggingLevel to the hidden value of "7" (change it back to "Debug" or lower or performance could be impacted) and look at the "C/C++" logging pane to see what folders are being iterated through and add those directories or symlinks that target those to C_Cpp.files.exclude . Look for the update_file_if_needed messages.

**Reply by whatdoidonow** [reply]
There was a holiday in the UK. So I've only just seen that and have yet to try this out.

**Reply by whatdoidonow** [reply]
I extended the logging level. Possibly simply because I was looking, but anyway, I saw these lines at the top of the output:
```
Attempting to get defaults from C++ compiler in "compilerPath" property: '/usr/bin/gcc'
  Folder: / will be indexed
Populate include completion cache.
Discovering files...
  Processing folder (recursive): /
process_path: /
```
That looks to me as a deliberate plan to scan the entire file system. No links involved. Oh, yes, I also scanned for links under the vscode workspace and there was only one innocuous link to a file. So the question is: where from did cpptools get into its head to scan / ? I probably should have said, I am on Ubuntu 18.04 LTS.

**Reply by sean-mcmanus** [reply]
Are you setting / to be the includePath? Or maybe querying /usr/bin/gcc is getting returning /? If you set compilerPath to "" that could eliminate that as the cause.

**Reply by whatdoidonow** [reply]
I very much appreciate you spending time to help me.
Having this line
```
"C_Cpp.default.compilerPath": ""
```
in the settings.json file stopped cpptools in its tracks. Meaning I couldn't spot it in the top running processes. So, if it runs, it's doing its job very quickly. Unfortunately, to run the debugger in vscode, I had to remove that line. But I guess the experiment gives a clue to the root cause. Not that I am very much the wiser, as I thought my gcc was the default install you get with Ubuntu.

**Reply by Colengms** [reply]
Hi @whatdoidonow .  Immediately after the first display of `Attempting to get defaults from C++ compiler in "compilerPath" property: '/usr/bin/gcc'`, the log should indicate a command line use to invoke the compiler to extract system include paths and system defines.  Normally, the list of detected system include paths would then be listed instead of "/".  Could you try running that logged command line manually and looking at what it returns?  Perhaps there are args being passed to it (maybe from a `compilerArgs` setting?) that are causing it not to query the compiler properly?

**Reply by whatdoidonow** [reply]
Thanks @Colengms . The majority of lines ```Attempting to get defaults from C++ compiler``` are not followed by a command. Those I found were:
"/usr/bin/gcc" -x c++ -E -dM /dev/null
-> a bunch of #defines with no paths, such as: #define \_\_SIZEOF_LONG_LONG__ 8

"/usr/bin/gcc" -x c -E -dM /dev/null
-> similar output

"/usr/bin/gcc" -dumpmachine
-> x86_64-linux-gnu

"/usr/bin/gcc" -std=gnu11 -Wp,-v -E -dD -x c -m64  /dev/null
-> I'll try and attach that output as c_includes.txt, but filtering for paths, I see
 /usr/lib/gcc/x86_64-linux-gnu/7/include
 /usr/local/include
 /usr/lib/gcc/x86_64-linux-gnu/7/include-fixed
 /usr/include/x86_64-linux-gnu
 /usr/include
[c_includes.stdout.txt](https://github.com/microsoft/vscode-cpptools/files/6590525/c_includes.stdout.txt)
[c_includes.txt](https://github.com/microsoft/vscode-cpptools/files/6590527/c_includes.txt)

"/usr/bin/gcc" -std=gnu++14 -Wp,-v -E -dD -x c++ -m64  /dev/null /usr/include/c++/7
-> I'll try and attach the output as cpp_includes.txt, but filtering for paths, I see
 /usr/include/x86_64-linux-gnu/c++/7
 /usr/include/c++/7/backward
 /usr/lib/gcc/x86_64-linux-gnu/7/include
 /usr/local/include
 /usr/lib/gcc/x86_64-linux-gnu/7/include-fixed
 /usr/include/x86_64-linux-gnu
 /usr/include
[cpp_includes.stdout.txt](https://github.com/microsoft/vscode-cpptools/files/6590528/cpp_includes.stdout.txt)
[cpp_includes.txt](https://github.com/microsoft/vscode-cpptools/files/6590530/cpp_includes.txt)

The .stdout.txt only collected stdout. The other two are combined stdout and stderr.
I am curious about the lines that did not list a command. Also, some of the commands were followed by
```terminating child process: 19454``` (varying pid, of course), as if it hadn't waited for closure. I might as well dump the start of the output window while I'm at it.
[start_of_debug_output.txt](https://github.com/microsoft/vscode-cpptools/files/6590629/start_of_debug_output.txt)

**Reply by Colengms** [reply]
Hi @whatdoidonow .  This appears to be the command used to query the compiler for configuration:

```
"/usr/bin/gcc" -std=gnu11 -Wp,-v -E -dD -x c -m64 /dev/null
```

The other two are detecting defaults based on defines (for C and C++, separately), and the default target (-dumpmachine).

Is it possible there is a symlink to root in one of the directory returned by the compiler?  That looks like what is happening.  Because the root is determined to encompass the other system include paths, only the root path is reported after the compiler query.

**Reply by whatdoidonow** [reply]
Apparently not. But I could have missed something and will search recursively before Monday. It has to be lurking somewhere.

**Reply by sean-mcmanus** [reply]
@Colengms  The message "Folder: / will be indexed" doesn't occur in response to directory iteration going into a symlink. It's only populated from the various settings that can set the browse.path...maybe we could add logging that shows where exactly the path is coming from. Might be better to just check for "/" and remove it.

**Reply by sean-mcmanus** [reply]
I filed bug https://github.com/microsoft/vscode-cpptools/issues/7648 -- the fix for that should solve your issue, even though I'm not sure what setting is exactly creating the "/" (maybe some C_Cpp.default.browse.path in the user settings.json?).

**Reply by whatdoidonow** [reply]
Just to put the link idea to bed: I determined that there are links to "/" on my system.
C_Cpp.default.browse.path isn't set. I couldn't see anything suspicious in the settings.json.
I'll take the suppression of "/" as an index target gratefully. But, of course, it is a little unsatisfying not knowing why it needs doing. If you need me to install a special logger, by all means, give me directions in due course.

**Reply by sean-mcmanus** [reply]
Is C_Cpp.default.includePath or C_Cpp.default.systemIncludePath set? When "/" is filtered out, it should be replaced by the other subdirectories that were referenced prior to the "/" removing them. I'm not sure of a good place to add more logging since I don't know where it would be useful.

**Reply by whatdoidonow** [reply]
C_Cpp.default.includePath and C_Cpp.default.systemIncludePath are not set in my settings.json.
I'm sorry I can't be more helpful.
Things are better for me, however, since I changed C_Cpp.workspaceParsingPriority to low. (Or maybe something else has changed?)
Many thanks to the contributors.

**Reply by sean-mcmanus** [reply]
With https://github.com/microsoft/vscode-cpptools/releases/tag/1.5.0-insiders we've added fix to filter out the root from directory iteration.

**Reply by whatdoidonow** [reply]
Thanks. I shall join the insiders as soon as possible.

**Reply by whatdoidonow** [reply]
```Discovering files: 28489 file(s) processed```
Seems reasonable to me (previously, I had 1.2 million being parsed).
Still took an hour and a bit for cpptools to settle down (remember I use "C_Cpp.workspaceParsingPriority": "low"). But that's okay.
So I'm happy for my issue to be closed. Thanks.
(I'd be happier still if we had discovered the weirdness in my system, but hey, I'm not going to be greedy.)
